src/org/sparkboard/server/server.clj

```diff
@@ -523,13 +390,213 @@
     (case action
       :link-account! (link-account! context params))))
 
+(defn action-value [action]
+  (case (:type action)
+    "checkboxes" (->> (:selected_options action)
+                      (map :value)
+                      (into #{}))
+    "static_select" (-> action :selected_option :value)
+    "plain_text_input" (:value action)
+    "button" nil
+    (do
+      (log/warn :not-parsing-action action)
+      action)))
+
+(defn actions-values [actions]
+  (into {} (map (juxt :action_id action-value)) actions))
+
+(defn view-values [view]
+  (->> (apply merge (vals (:values (:state view))))
+       (reduce-kv (fn [m k action]
+                    (assoc m k (action-value action))) {})
+       (merge (:private_metadata view))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Testing local-state
+
+(defn modal-updater
+  "Returns a handler that will update a modal, given:
+   - modal-fn, accepts [context, payload, state] & returnsn a [:modal ..] hiccup form
+   - reducer, a function of [state] that returns a new state"
+  [modal-fn reducer]
+  (fn [k context {:keys [view actions] :as payload} {:keys [modal!]}]
+    (let [next-state (reducer (merge (:private_metadata view) (actions-values actions)))]
+      (modal! (:id view)
+              (-> (modal-fn context payload next-state)
+                  (assoc-in [1 :private_metadata] next-state))))))
+
+(defn modal-opener
+  "Returns a handler that will open a modal, given:
+   - modal-fn, accepts [context, payload, state] & returnsn a [:modal ..] hiccup form
+   - initial-state, a map"
+  [modal-fn initial-state]
+  (fn [_ context payload {:keys [modal!]}]
+    (modal! (-> (modal-fn context payload initial-state)
+                (assoc-in [1 :private_metadata] initial-state)))))
+
+;; TODO
+;; a modal 'builder' function that will return all of the handlers
+;; needed to support a modal + its local state updates
+(comment
+  (modal-builder "my-modal"
+                 {:initial-state {:n 0}}
+                 (fn [state]
+                   [:modal {:title "Hello"}
+                    [:actions
+                     [:button {:modal/action-id "counter+"
+                               :modal/on-action #(update % :n inc)}
+                      (str "Counter: " (:n state))]]]))
+  ;; this would return
+  {:block_actions/my-modal-open ...
+   :block_actions/my-modal-counter+ ...})
+;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(def slack-api-handlers
```

Comment:
Putting events in one place (but distinguishable by namespace) works quite well here

Reply:
I am thinking so too. I started with a nested structure but it's much easier to work with this as a flat map.